src/VisualStudio/Core/Test.Next/Services/ServiceHubServicesTests.cs

```diff
@@ -321,6 +327,348 @@ public async Task TestRemoteWorkspaceCircularReferences()
+            DoPermute(0, values.Length - 1);
+            return result.ToImmutable();
+
+            void DoPermute(int start, int end)
+            {
+                if (start == end)
+                {
+                    // We have one of our possible n! solutions,
+                    // add it to the list.
+                    result.Add(values.ToImmutableArray());
+                }
+                else
+                {
+                    for (var i = start; i <= end; i++)
+                    {
+                        (values[start], values[i]) = (values[i], values[start]);
+                        DoPermute(start + 1, end);
+                        (values[start], values[i]) = (values[i], values[start]);
+                    }
+                }
+            }
+        }
+
+        private static async Task TestInProcAndRemoteWorkspace(
+            bool syncWithRemoteServer, params ImmutableArray<(string hintName, SourceText text)>[] values)
+        {
+            // Try every permutation of these values.
+            foreach (var permutation in Permute(values))
+            {
+                var gotException = false;
+                try
+                {
+                    await TestInProcAndRemoteWorkspaceWorker(syncWithRemoteServer, permutation);
+                }
+                catch (XunitException)
+                {
+                    gotException = true;
+                }
+
+                // If we're syncing to the remove server, we should get no exceptions, since the data should be matched
+                // between both.  If we're not syncing, we should see a failure since the two processes will disagree on
+                // the contents.
+                Assert.Equal(syncWithRemoteServer, !gotException);
+            }
+        }
+
+        [ExportWorkspaceService(typeof(IWorkspaceConfigurationService), ServiceLayer.Test), SharedAttribute, PartNotDiscoverable]
+        private sealed class NoSyncWorkspaceConfigurationService : IWorkspaceConfigurationService
+        {
+            [ImportingConstructor]
+            [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+            public NoSyncWorkspaceConfigurationService()
+            {
+            }
+
+            public WorkspaceConfigurationOptions Options => WorkspaceConfigurationOptions.Default with { RunSourceGeneratorsInProcessOnly = true };
+        }
+
+        private static async Task TestInProcAndRemoteWorkspaceWorker(
+            bool syncWithRemoteServer, ImmutableArray<ImmutableArray<(string hintName, SourceText text)>> values)
+        {
+            var throwIfCalled = false;
+            ImmutableArray<(string hintName, SourceText text)> sourceTexts = default;
+            var generator = new CallbackGenerator(
+                onInit: _ => { },
+                onExecute: _ => { },
+                computeSourceTexts: () =>
+                {
+                    Contract.ThrowIfTrue(throwIfCalled);
```

Comment:
If this throws, is this going to get caught by the driver and not propogated to your test as you might expect?

Reply:
yes.  but that's ok.  we then get assertions later because the doc content checks fail :)